third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,248 @@
+
+    auto srcType = cvtOp.getSrc().getType();
+    auto dstType = cvtOp.getType();
+
+    auto srcEnc = srcType.getEncoding();
+    auto dstEnc = dstType.getEncoding();
+
+    auto ctx = srcEnc.getContext();
+    auto rank = srcType.getShape().size();
+    unsigned numWarps = triton::gpu::getNumWarpsPerCTA(srcEnc);
+    auto warpSize = triton::gpu::getWarpSize(srcEnc);
+
+    // Find all possible shapes of WarpsPerCTA by finding all possible
+    // factorizations of numWarps. Pick shape for which both conversions in
+    // decomposition use LDS less than LDSSize and for which sum of LDS usage
+    // is minimal. If no such shape exists, do not decompose.
+    auto factorizedNumWarps =
+        mlir::triton::AMD::factorizePowerOf2(numWarps, rank);
+    // Create a list of temporary layouts
+    SmallVector<Attribute> tmpLayouts;
+    for (int i = 0; i < factorizedNumWarps.size(); i++) {
+      auto warpsPerCTA = factorizedNumWarps[i];
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(srcEnc, warpsPerCTA));
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(dstEnc, warpsPerCTA));
+      SmallVector<unsigned> elemsPerThread(rank, 1);
+      SmallVector<unsigned> threadsPerWarp(rank, 1);
+      threadsPerWarp[rank - 1] = warpSize / 8;
+      threadsPerWarp[rank - 2] = warpSize / threadsPerWarp[rank - 1];
+      auto order = triton::gpu::getOrder(srcEnc);
+      auto layoutCTA = triton::gpu::getCTALayout(srcEnc);
+      auto fallbackLayout = triton::gpu::BlockedEncodingAttr::get(
+          ctx, elemsPerThread, threadsPerWarp, warpsPerCTA, order, layoutCTA);
+      tmpLayouts.push_back(fallbackLayout);
+    }
+
+    unsigned minLDSUsage = 2 * LDSSize;
+    int minIdx = -1;
+    for (int i = 0; i < tmpLayouts.size(); i++) {
+      auto resources = mlir::triton::AMD::estimateResourcesForReplacement(
+          builder, cvtOp, tmpLayouts[i]);
+      // TODO analyze performance along with LDS consumption
+      if (resources.LDS < minLDSUsage) {
+        minLDSUsage = resources.LDS;
+        minIdx = i;
+      }
+    }
+
+    if (minIdx == -1 || minLDSUsage > targetLDSSize) {
+      return;
+    }
+
+    triton::gpu::ConvertLayoutOp tmpCvt;
+    triton::gpu::ConvertLayoutOp newEpilogueCvt;
+
+    assert(minIdx >= 0 && minIdx < tmpLayouts.size());
+    auto tmpLayout = tmpLayouts[minIdx];
+    std::tie(tmpCvt, newEpilogueCvt) =
+        mlir::triton::AMD::createNewConvertOps(builder, cvtOp, tmpLayout);
+
+    cvtOp.replaceAllUsesWith(newEpilogueCvt.getResult());
+    cvtOp.erase();
+  }
+
+  struct LDSBottleneckOperation {
+    triton::gpu::ConvertLayoutOp op;
+    int64_t LDSSizeTarget;
+  };
+
+  /**
+   * Assuming that all buffer above scratch buffer in memory space can be
+   * shifted down in memory, this function gives optimistic estimation of memory
+   * space available for scratch buffer.
+   */
+  int64_t
+  computeMaxScratchBufferSize(triton::gpu::ConvertLayoutOp op,
```

Comment:
Maybe `computeTargetBufferSize`? I feel like "target" or "desired" is more accurate about what we want to do here.